### DISCUSSION TITLE
`resource` template `merge` & `strategic` patch error about using both `-f` and `-p`

### DISCUSSION THREAD

**Discussion by djquiambao18** [argoproj/argo-workflows#13279] [Q&A] [answered]
Hello,
in https://github.com/argoproj/argo-workflows/blob/06da23e8660f7841592070b1a372ace0750c2364/workflow/executor/resource.go#L137

the `-f` flag is not added for `mergeStrategy: json` but any others add this per the `appendFileFlag`. In the comments preceding Line 137, it seems that the implementer was aware of this limitation.

However, just a few lines down: https://github.com/argoproj/argo-workflows/blob/06da23e8660f7841592070b1a372ace0750c2364/workflow/executor/resource.go#L144
the `-p` is still added along with the `-f` flag later on, inevitably causing the "patch" action to be broken.

For instance, when you try to use patch action with resource something like this example: https://github.com/argoproj/argo-workflows/blob/main/examples/k8s-patch.yaml - (my use-case is with ConfigMap)
 
the engine parses it into a `kubectl` command like:
```
level=info msg=kubectl patch --type strategic -p '{\"data\": {\"azure-dev\": \"true\"}}'\n configmap deploy-status -f /tmp/manifes
t.yaml -o json
```
which will give an error:
```
level=error msg="executor error: no more retries error: when paths, URLs, or stdin is provided as input, you may not specify resource arguments as well"
```

Is this the intended behavior and am I simply misunderstanding something here? It seems that the only option to prevent this issue is to use the `json` mergeStrategy at the moment.

Thanks and clarification is appreciated!

**Comment by agilgur5** [comment]
It looks like someone else made a PR to attempt to change this, but they realized it was incorrect: https://github.com/argoproj/argo-workflows/pull/11249#issuecomment-1600745498:

> The current way to do it is to write instead
> 
> ```yaml
>       resource:
>         action: patch
>         mergeStrategy: merge
>         manifest: |
>           apiVersion: v1
>           kind: ConfigMap
>           metadata:
>             name: name-of-configmap
>           data:
>             test: "Goodbye"
> ```

**Reply by djquiambao18** [reply]
Sorry, I didn't mention that I was aware that the above works, but it felt more like a workaround than actually intended, with the json patch design in mind. I guess my expectation was that with kubectl commands, I didn't have to specify the manifest for merge or strategic patching and could do in-line patch like `kubectl patch cm <cm-name> -p '{"data": {"test" : "Goodbye"}}'`
and equivalently in the workflow (default patch type strategic):
```yaml
      resource:
        action: patch
        manifest: |
          {"data": {"test" : "Goodbye"}}
```

the nice thing about not having to use json patch and the above example you mentioned is that the key "data" didn't have to exist beforehand and I didn't have to know the metadata of the configmap beforehand other than its name.

**Reply by agilgur5** [reply]
Mm is the problem that `merge` and `strategic` accept both a full manifest _and_ a partial one?
As such the spec doesn't know which one you're trying to use?

If so, I guess we might be able to infer a partial based on the `manifest` (e.g. lack of `metadata.name` etc) or other `flags`. PR welcome

**Reply by agilgur5** [reply]
Hmm, we might just be able to use `--patch-file` universally for all patches instead of `-f`

**Reply by agilgur5** [reply]
Threw up a draft PR of that^: #13317. Still needs some testing though

**Reply by agilgur5** [reply, marked as answer]
Fixed in #13317. 
This discussion also effectively ends up as a duplicate of the now fixed #11248 

More details:
`--patch-file` universally doesn't quite work and has the same problem as `-p`.

When a full manifest is given, effectively _both_ `-f` and `--patch-file` are needed, pointing to the same manifest. `-f` gets the `kind` and `name` from the manifest (the "header") and `--patch-file` uses the manifest as a patch file.

So I instead used the strategy I outlined [above](https://github.com/argoproj/argo-workflows/discussions/13279#discussioncomment-9941246) to infer a "full" manifest with a header by inspecting if it has a `kind`.
